src/Microsoft.Identity.Client/MsalServiceException.cs

```diff
@@ -246,5 +216,36 @@ internal override void PopulateObjectFromJson(JObject jobj)
             CorrelationId = JsonUtils.GetExistingOrEmptyString(jobj, CorrelationIdKey);
             SubError = JsonUtils.GetExistingOrEmptyString(jobj, SubErrorKey);
         }
+
+        /// <summary>
+        /// Classification of the conditional access error, enabling you to do more actions or inform the user depending on your scenario. See https://aka.ms/msal-net-uiexception for details.
+        /// </summary>
+        /// <remarks>This class <see cref="MsalUiRequiredException"/> lists most classification strings as constants. </remarks>
+        public string InvalidGrantClassification
+        {
+            get
+            {
+                switch (SubError)
+                {
+                    case Client.InvalidGrantClassification.BasicAction:
+                    case Client.InvalidGrantClassification.AdditionalAction:
+                    case Client.InvalidGrantClassification.MessageOnly:
+                    case Client.InvalidGrantClassification.ConsentRequired:
+                    case Client.InvalidGrantClassification.UserPasswordExpired:
+                        return SubError;
+
+                    case Client.InvalidGrantClassification.BadToken:
+                    case Client.InvalidGrantClassification.TokenExpired:
+                    case Client.InvalidGrantClassification.ProtectionPolicyRequired:
+                    case Client.InvalidGrantClassification.ClientMismatch:
+                    case Client.InvalidGrantClassification.DeviceAuthenticationFailed:
+                        return string.Empty;
+
+                    // Forward compatibility - new sub-errors bubble through
+                    default:
+                        return SubError;
```

Comment:
perhaps we should flip this to only expose what we know? Perhaps there will be more like the above where we don't expose the invalid grant classification... not sure. Alternatively we can expose everything....

Reply:
to be considered. currently leaning towards not exposing it and not causing a ui require exception by default

---
In reply to: [298366905](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1240#discussion_r298366905) [](ancestors = 298366905)